lib/application_insights/rack/track_request.rb

```diff
@@ -19,42 +20,53 @@ def initialize(app, instrumentation_key, buffer_size = 500, send_interval = 60)
         @instrumentation_key = instrumentation_key
         @buffer_size = buffer_size
         @send_interval = send_interval
+
+        @sender = Channel::AsynchronousSender.new
+        @sender.send_interval = @send_interval
+        queue = Channel::AsynchronousQueue.new @sender
+        queue.max_queue_length = @buffer_size
+        @channel = Channel::TelemetryChannel.new nil, queue
+
+        @client = TelemetryClient.new @instrumentation_key, @channel
       end
 
       # Track requests and send data to Application Insights asynchronously.
       # @param [Hash] env the rack environment.
       def call(env)
+        # Create a duplicate of the middleware for each request to ensure thread safety
+        dup._call(env)
+      end
+
+      # Track requests and send data to Application Insights asynchronously.
+      # @param [Hash] env the rack environment.
+      def _call(env)
+        # Build a request ID, incorporating one from our request if one exists.
+        request_id = request_id_header(env['HTTP_REQUEST_ID'])
+        env['ApplicationInsights.request.id'] = request_id
+
         start = Time.now
         begin
           status, headers, response = @app.call(env)
+
+          # Set a Request-Id response header if one has not been set within our app
```

Comment:
@yantang-msft this is to keep in line with the specifications mentioned above?

It states somewhere that you should expose the Request-Id in the instance where the application using ApplicationInsights-Ruby is service-b?

I may have mis-understood?

Reply:
@mattrayner , could you give the detailed specification for this?
> you should expose the Request-Id in the instance where the application using ApplicationInsights-Ruby is service-b

Supposing this method is called on service-b. ```env[Http_Request_Id]``` is the id for the outgoing request on service-a. Then you generate the ```request-id```, which is the id of the incoming request on service-b. And it's stored in ```env[ApplicationInsights.request.id]```. I guess ```env[ApplicationInsights.request.id]``` is what you meant by 'expose id in the instance'? Then it's already exposed, no need to set the Request-Id in response header.

Reply:
It may be that I've mis understood the implementation but `headers['Request-Id']` refers to the response header that will be sent back to service-a, detailing service-b's Request-Id?

Reply:
I feel like I must have mis-understood - I'll remove this

Reply:
The request-id is sent only from upstream to downstream. But I just looked at the document and also get a bit confused by the example [here](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/HierarchicalRequestId.md#example), which makes me think I need to pass the request-id of service-b through Correlation-Context in the response header.

@lmolkova , could you help clarifying whether it is a problem in the document? I don't remember we set correlation-context or request-id in the response header in other sdks.